server/src/test/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplateTests.java

```diff
@@ -240,18 +242,85 @@ public void testXContentSerializationWithRolloverAndEffectiveRetention() throws
             RolloverConfiguration rolloverConfiguration = RolloverConfigurationTests.randomRolloverConditions();
             DataStreamGlobalRetention globalRetention = DataStreamGlobalRetentionTests.randomGlobalRetention();
             ToXContent.Params withEffectiveRetention = new ToXContent.MapParams(DataStreamLifecycle.INCLUDE_EFFECTIVE_RETENTION_PARAMS);
-            template.toXContent(builder, withEffectiveRetention, rolloverConfiguration, globalRetention);
+            // non-system data stream gets effective_retention:
+            template.toXContent(builder, withEffectiveRetention, rolloverConfiguration, globalRetention, false);
             String serialized = Strings.toString(builder);
             assertThat(serialized, containsString("rollover"));
-            for (String label : rolloverConfiguration.resolveRolloverConditions(lifecycle.getEffectiveDataRetention(globalRetention))
-                .getConditions()
-                .keySet()) {
+            for (String label : rolloverConfiguration.resolveRolloverConditions(
+                lifecycle.getEffectiveDataRetention(globalRetention, randomBoolean())
```

Comment:
Shouldn't the last parameter here match what you pass to `template.toXContent` a couple of lines higher? So, `false` instead of `randomBoolean`. The same applies to some other tests you changed.